packages/esm-patient-vitals-app/src/vitals-biometrics-form/vitals-biometrics-form.component.tsx

```diff
@@ -87,22 +62,152 @@ const VitalsAndBiometricsForm: React.FC<DefaultWorkspaceProps> = ({
-  const [hasInvalidVitals, setHasInvalidVitals] = useState(false);
   const [isSubmitting, setIsSubmitting] = useState(false);
   const [muacColorCode, setMuacColorCode] = useState('');
-  const [showErrorNotification, setShowErrorNotification] = useState(false);
-  const [showErrorMessage, setShowErrorMessage] = useState(false);
+  const [showFormSubmissionErrorNotifications, setShowFormSubmissionErrorNotifications] = useState(false);
+
+  const concepts = useMemo(
+    () => ({
+      midUpperArmCircumference: conceptRanges.get(config.concepts.midUpperArmCircumferenceUuid),
+      diastolicBloodPressure: conceptRanges.get(config.concepts.diastolicBloodPressureUuid),
+      systolicBloodPressure: conceptRanges.get(config.concepts.systolicBloodPressureUuid),
+      oxygenSaturation: conceptRanges.get(config.concepts.oxygenSaturationUuid),
+      respiratoryRate: conceptRanges.get(config.concepts.respiratoryRateUuid),
+      temperature: conceptRanges.get(config.concepts.temperatureUuid),
+      weight: conceptRanges.get(config.concepts.weightUuid),
+      height: conceptRanges.get(config.concepts.heightUuid),
+      pulse: conceptRanges.get(config.concepts.pulseUuid),
+    }),
+    [
+      conceptRanges,
+      config.concepts.diastolicBloodPressureUuid,
+      config.concepts.heightUuid,
+      config.concepts.midUpperArmCircumferenceUuid,
+      config.concepts.oxygenSaturationUuid,
+      config.concepts.pulseUuid,
+      config.concepts.respiratoryRateUuid,
+      config.concepts.systolicBloodPressureUuid,
+      config.concepts.temperatureUuid,
+      config.concepts.weightUuid,
+    ],
+  );
+
+  const decimalRefinement = (key: keyof VitalsBiometricsFormData, schema: z.ZodNumber) => {
+    const decimalCountAllowed = getDecimalCountForField(key);
+    return schema.refine(
+      (value) => {
+        if (value === undefined) {
+          return true;
+        }
+        if (!decimalCountAllowed) {
+          return Number.isInteger(value);
+        }
+
+        const decimalCount = value.toString().split('.')?.[1]?.length ?? 0;
+        return decimalCount <= decimalCountAllowed;
+      },
+      decimalCountAllowed
+        ? t('limitedDecimalPlacesAllowed', 'Value must be a number with {{count}} decimal places', {
+            count: decimalCountAllowed,
+          })
+        : t('valueMustBeInteger', 'Value must be an integer'),
+    );
+  };
+
+  const getZodNumberSchema = (key: keyof VitalsBiometricsFormData, min: number, max: number) => {
+    let numberSchema = z.number();
+    if (min) {
+      numberSchema = numberSchema.min(
+        min,
+        t('validationInputError', 'Value must be between {{min}} and {{max}}', { min, max }),
+      );
+    }
+    if (max) {
+      numberSchema = numberSchema.max(
+        max,
+        t('validationInputError', 'Value must be between {{min}} and {{max}}', { min, max }),
+      );
+    }
+    return decimalRefinement(key, numberSchema);
+  };
+
+  const VitalsAndBiometricFormSchema = z
```

Comment:
If possible, `decimalRefinement()`, `getZodNumberSchema()` and this should be refactored to be outside the component. If not, they should be wrapped in `useCallback()` and `useMemo()`, because as it stands, `VitalsAndBiometricFormSchema` will be a different object on each render, which may cause a re-render cascade.